### PR TITLE
[FrameworkBundle] Fix a regression in handling absolute template paths

### PR DESCRIPTION
**Pull request by jakzal**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17777 #17683 |
| License | MIT |
| Doc PR | - |

Regression introduced by #15272.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php [hunk 1]
```diff
@@ -56,7 +56,7 @@ public function parse($name)
             throw new \RuntimeException(sprintf('Template name "%s" contains invalid characters.', $name));
         }
 
-        if (!preg_match('/^(?:([^:]*):)?(?:([^:]*):)?(.+)\.([^\.]+)\.([^\.]+)$/', $name, $matches)) {
+        if (!preg_match('/^(?:([^:]*):([^:]*):)?(.+)\.([^\.]+)\.([^\.]+)$/', $name, $matches) || $this->isAbsolute($name)) {
```
Comment (jakzal) [inline src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php]: Two groups replaced with one here. This will only match "Foo:Bar:" or "".

**Comment by jakzal** [comment]
@francoispluchino @wesleylancel would you mind checking if this fixes your issues please?

**Comment by francoispluchino** [comment]
@jakzal It's OK for me! Thank you for this PR.

**Comment by jakzal** [comment]
I just confirmed it also fixes the issue reported by @Congelli501. Waiting for @wesleylancel to either confirm his case's fixed, or to provide a failing scenario.

**Comment by fabpot** [comment]
Not really related to this PR, but loading templates via an absolute path looks terrible.

**Comment by francoispluchino** [comment]
@fabpot I agree with you if you load  a template from a Bundle or App Resources, but when the template file is generated dynamically, and it is not in a Bundle we have no other choice.

Personally, I have this problem with the mail templates that are accessible via a custom stream, with the email generator that provided me the absolute path of the template. So, dynamically add the template file before loading the template is not very subtle, Because the `templating.locator` will search the template, while the path is already known ...

Of course, I am interested by best practices, if you have any.

**Comment by jakzal** [comment]
Back to the topic... I confirmed wesleylancel's problem still persists. I'll work on a fix this afternoon.

**Comment by jakzal** [comment]
@hcomnetworkers can you try this patch please? 

status: needs review

**Comment by hcomnetworkers** [comment]
@jakzal I have tested your patch and it works for #17958, thank you!

**Comment by jakzal** [comment]
Several people confirmed this fixes the issue for them.

status: reviewed

File: src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php [hunk 2]
```diff
@@ -72,4 +72,9 @@ public function parse($name)
 
         return $this->cache[$name] = $template;
     }
+
+    private function isAbsolute($path)
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php]: Can you rename to `isAbsolutePath()` as this is the name we are using elsewhere.

Also, I'd prefer to use a more restrictive approach like done elsewhere:

```php
    private function isAbsolutePath($file)
    {
        return $file[0] === '/' || $file[0] === '\\'
            || (strlen($file) > 3 && ctype_alpha($file[0])
                && $file[1] === ':'
                && ($file[2] === '\\' || $file[2] === '/')
            );
    }

```

Comment (jakzal) [inline src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php]: This won't work with windows paths which are "normalised" for some reasons here: 

```php
$name = str_replace(':/', ':', preg_replace('#/{2,}#', '/', str_replace('\\', '/', $name)));
```

**Comment by fabpot** [comment]
Thank you @jakzal.

**Comment by stof** [comment]
> but when the template file is generated dynamically, and it is not in a Bundle we have no other choice.

The right way to do this would be to come up with your own convention about template names and a template loader able to resolve it. See https://github.com/Behat/Borg/blob/2d8422d006b32cff929a0e475acb473258911baf/src/Integration/Symfony/Documentation/Twig/DocumentationLoader.php for an example doing it

@fabpot @jakzal what do you think about deprecating the possibility to use absolute paths ? AFAIK, this was never intended to work. It was just a side-effect of the initial implementation.

**Comment by fabpot** [comment]
@stof Indeed, it works by chance. We should definitely deprecate this. see #17999

**Comment by jakzal** [comment]
@stof indeed looks like many cases were supported unintentionally. Since there was no test cases for them I wasn't sure if it was intended. I'll send a PR to deprecated absolute paths.

File: src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php [hunk 3]
```diff
@@ -56,7 +56,7 @@ public function parse($name)
             throw new \RuntimeException(sprintf('Template name "%s" contains invalid characters.', $name));
         }
 
-        if (!preg_match('/^(?:([^:]*):)?(?:([^:]*):)?(.+)\.([^\.]+)\.([^\.]+)$/', $name, $matches)) {
```
Comment (chrjean) [inline src/Symfony/Bundle/FrameworkBundle/Templating/TemplateNameParser.php]: That breaks AppBundle:Some/path pattern

**Pull request merged**